### DISCUSSION TITLE
[Proposal]: Add explicit self-referencing type constraint

### DISCUSSION THREAD

**Discussion by huoshan12345** [dotnet/csharplang#5126] [Language Ideas]
# FEATURE_NAME

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
Currently, if we want to define a generic class with self-referencing type constraint, we can do this:
```
public interface ISingleton<TSelf> where TSelf : ISingleton<TSelf>
{
    static abstract TSelf Instance { get; }
}
public class SingletonA : ISingleton<SingletonA>
{
    public static SingletonA Instance { get; } = new SingletonA();
}
```
It works well, but there is one unexpected usage:  
```
public class SingletonB : ISingleton<SingletonA>
{
    public static SingletonA Instance { get; } = new SingletonA();
}
```
The class `SingletonB` is defined legally but not reasonable cause `SingletonB` should implement `ISingleton<SingletonB>` instead of `ISingleton<SingletonA>`. It is a misuse of `ISingleton<TSelf>` and cannot be prevented.

## Motivation
To prevente the misuse of self-referencing type constraint.

## Detailed design
[design]: #detailed-design

Add a new keyword `self` to generic constraint, which restricts that the generic type must be the implement type itself.
So we can rewrite the above code as:
```
public interface ISingleton<TSelf> where TSelf : self
{
    static abstract TSelf Instance { get; }
}
public class SingletonA : ISingleton<SingletonA>
{
    public static SingletonA Instance { get; } = new SingletonA();
}
```
And the following is illegal now:
```
public class SingletonB : ISingleton<SingletonA> // compile error: type mismatch, should be ISingleton<SingletonB>
{
    ...
}
```

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by 333fred** [comment]
I'm converting this to a discussion, as outlined in the README. https://github.com/dotnet/csharplang#discussions

**Comment by Joe4evr** [comment]
You can already simulate this with analyzers, [and I've made a package for constraints including this](https://www.nuget.org/packages/JiiLib.Constraints/).
```cs
using JiiLib.Constraints;

public interface ISingleton<[SelfType] TSelf> where TSelf : ISingleton<TSelf>
{
    static abstract TSelf Instance { get; }
}
```

**Comment by sodiboo** [comment]
I think this is mostly a nonissue because having this interface appear directly in a parameter/field/variable type doesn't make much sense, and it's much more likely to be a constraint against a generic T, such as [``TSelf Create<TOther>(TOther value) where TOther : INumber<TOther>``](https://github.com/dotnet/runtime/blob/7c4e6061e75b1400a2b7b9ad3739d643a8e4ea57/src/libraries/System.Private.CoreLib/src/System/INumber.cs#L54-L77) (from INumber, of course ignore the names of everything because they confuse my point here but i'm giving a real world example). For such a type, ``SingletonB`` would not be a valid constraint because, although it implements some ``ISingleton``, it does not implement ``ISingleton<SignletonB>``. And this also makes sense for your case, because for generic static interface constraints, you could have a method something like this:

```cs
T GetInstanceOf<T>() where T : ISingleton<T> {
    return T.Instance;
}

var a = GetInstanceOf<SingletonA>(); // perfectly ok, and this type is the one that implements it as expected
var b = GetInstanceOf<SingletonB>(); // error CS0311: The type 'SingletonB' cannot be used as type parameter 'T' in the generic type or method 'GetInstanceOf<T>()'. There is no implicit reference conversion from 'SingletonB' to 'ISingleton<SingletonB>'.
```

Even though the declaration is fine, you do get an issue if you try to actually use ``SingletonB`` in any sensible location. And of course, as mentioned, if you want an error when declaring such an invalid use of that type, there are always analyzers that can help you detect that.

**Comment by msedi** [comment]
Wouldn't it be helpful then to also enable the following with a self keyword?

```cs
public abstract class Car
{
  public abstract self CreateCopy();
}

public class Audi : Car
{
  public override self CreateCopy();
}
```
